contrib/babelfishpg_tsql/src/collation.c

```diff
@@ -450,16 +451,160 @@ transform_from_ci_as_for_likenode(Node *node, OpExpr *op, like_ilike_info_t like
+		return;
+	}
+#endif
+	elog(LOG, "Using ICU function to remove accents");
 	remove_accents_internal_oid = LookupFuncName(list_make2(makeString("sys"), makeString("remove_accents_internal")), -1, funcargtypes, true);
 }
 
+/*
+ * store 32bit character representation into multibyte stream
+ */
+static inline void
+store_coded_char(unsigned char *dest, uint32 code)
+{
+	if (code & 0xff000000)
+	{
+		*dest++ = code >> 24;
+	}
+	if (code & 0x00ff0000)
+	{
+		*dest++ = code >> 16;
+	}
+	if (code & 0x0000ff00)
+	{
+		*dest++ = code >> 8;
+	}
+	if (code & 0x000000ff)
+	{
+		*dest++ = code;
+	}
+	*dest = '\0';
+	return;
+}
+
+static int
+compare_remove_accent_map_pair(const void *p1, const void *p2)
+{
+	uint32		v1,
+				v2;
+
+	v1 = *(const uint32 *) p1;
+	v2 = ((const remove_accent_map_pair *) p2)->original_char;
+	return (v1 > v2) ? 1 : ((v1 == v2) ? 0 : -1);
+}
+
+PG_FUNCTION_INFO_V1(remove_accents_internal_using_cache);
+Datum remove_accents_internal_using_cache(PG_FUNCTION_ARGS)
+{
+	unsigned char *input_str = (unsigned char *) text_to_cstring(PG_GETARG_TEXT_PP(0));
+	unsigned char *input_str_start = input_str;
+	text          *return_result;
+	int           len = strlen((char *)input_str),
+	              l;
+	StringInfoData result;
+	unsigned char *c = (unsigned char *) palloc(sizeof(uint32) + 1);
+	initStringInfo(&result);
+
+	for (; len > 0; len -= l)
+	{
+		unsigned char b1 = 0;
+		unsigned char b2 = 0;
+		unsigned char b3 = 0;
+		unsigned char b4 = 0;
+		uint32 utf8_char;
+		uint32 utf8_normalized_str;
+		remove_accent_map_pair *pr;
+
+		/* "break" cases all represent errors */
+		if (*input_str == '\0')
+			break;
+
+		l = pg_utf_mblen(input_str);
+		
+		if (len < l)
+			break;
+
+		if (!pg_utf8_islegal(input_str, l))
+			break;
+
+		if (l == 1)
+		{
+			appendBinaryStringInfo(&result, input_str++, 1);
+			continue;
+		}
+
+		/* collect coded char of length l */
+		if (l == 2)
+		{
+			b3 = *input_str++;
+			b4 = *input_str++;
+		}
+		else if (l == 3)
+		{
+			b2 = *input_str++;
+			b3 = *input_str++;
+			b4 = *input_str++;
+		}
+		else if (l == 4)
+		{
+			b1 = *input_str++;
+			b2 = *input_str++;
+			b3 = *input_str++;
+			b4 = *input_str++;
+		}
+		else
+		{
+			elog(ERROR, "unsupported character length %d", l);
+		}
+
+		utf8_char = (b1 << 24 | b2 << 16 | b3 << 8 | b4);
+
+		pr = bsearch(&utf8_char, pltsql_remove_accent_map, lengthof(pltsql_remove_accent_map),
+							sizeof(remove_accent_map_pair), compare_remove_accent_map_pair);
+
+		/* Use the mapping if availaible or else the character */
+		if (pr && pr->normalized_char)
+			utf8_normalized_str = pr->normalized_char;
+		else
+			utf8_normalized_str = utf8_char;
+
+		store_coded_char(c, utf8_normalized_str);
+
+		appendBinaryStringInfo(&result, c, strlen((const char *) c));
+	}
+
+	if (len > 0)
+		ereport(ERROR,
+			(errcode(ERRCODE_CHARACTER_NOT_IN_REPERTOIRE),
+			 errmsg("invalid byte sequence for encoding UTF-8 while removing accents using cache")));
```

Comment:
let us not talk about cache, it is internal. Try to given operation context in which this error can happen

Reply:
fixed